### ISSUE TITLE
Add alternative, procedural transaction API

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#5376]
The existing transaction API is nice in most cases. But sometimes, a procedural approach is more desireable. The idea would be to have an API similar to JDBC or JTA:

``` java
TransactionalDSLContext tx = ctx.beginTransaction();
tx.insert()...
tx.update()...
Savepoint savepoint = tx.savepoint();
tx.delete()...
tx.commit();
```

This new API should integrate flawlessly with the existing API, and should adhere to the behaviour imposed by the `Configuration.transactionProvider()`

As a workaround, users can access JDBC directly, very easily:

```java
// Begin a transaction
ctx.connection(c -> {
    DSLContext ctx2 = ctx.configuration().derive(c);

    // Begin transaction
    boolean autoCommit = c.getAutoCommit();

    try {
        c.setAutoCommit(false);
        ctx2.insert()...
        ctx2.update()...
        Savepoint savepoint = c.setSavepoint();
        ctx2.delete()...
        c.commit();
    }
    finally {
        c.setAutoCommit(autoCommit);
    }
});
```

---

See also:
- https://groups.google.com/d/msg/jooq-user/1JwWMChD2SM/NHUhSnI8AgAJ
- http://stackoverflow.com/q/38134201/521799

**Comment by lukaseder** [comment]
This isn't urgent and probably rather tricky to implement correctly. Will postpone this. Users can always access the JDBC transaction API directly via `DSLContext.connection(ConnectionRunnable)`

**Comment by chuanhhoang** [comment]
Do we have to setAutoCommit(true) after we commit the transaction?

**Comment by lukaseder** [comment]
@chuanhhoang can you please create a new issue for your question? Please include more details. It isn't strictly related to this feature request

**Comment by salimane** [comment]
@lukaseder in the above example, is `c` a new connection? If I am doing the above in a paralleStream, am I creating new connections for each item?

**Comment by lukaseder** [comment]
Thanks for your message, @salimane. I think your question isn't related to "alternative, procedural transaction APIs." (i.e. this issue) It's an entirely new question. May I ask you to create a new issue for your question. https://github.com/jOOQ/jOOQ/issues/new/choose, or maybe even ask it on Stack Overflow for more reach: https://stackoverflow.com/questions/tagged/jooq

**Comment by lukaseder** [comment]
I've never been too much of a fan of this approach, having maintained tons of really messy procedurally transactional PL/SQL code in the past. It is never really very clear in what transactional state any given code is. With the transactional lambda API jOOQ is offering, things are always quite clear.

- We nest => we create a savepoint
- We throw => we roll back
- We commit by default

This model works well for blocking/async/reactive. The suggested procedural transaction API only works for blocking. We'd have to add weird overloads for async/reactive, too, all the while offering two separate APIs for the same thing, which are quite confusing, in addition to there being a third option: Using Spring or Java EE's declarative transactions, completely outside of jOOQ.

All the while it is super simple to write a custom wrapper around JDBC that takes care of *just that*, and *only that*. jOOQ's transaction API is opinionated. It uses lambdas. If that's not the right approach, then it's always possible to revert to JDBC or Spring.

I won't implement this.

**Comment by lukaseder** [comment]
An additional issue that was avoided by not implementing this is a potential API conflict with https://github.com/jOOQ/jOOQ/issues/7106, where `COMMIT`, `ROLLBACK`, `SAVEPOINT` procedural statements might be supported.

**Comment by cowwoc** [comment]
But... but... Loom! :)

On a more serious note, virtual threads remove the need for async/reactive models and they fix the problem with lambdas not being able to throw checked exceptions, access non-final parameters, and are much easier to read, debug, and so on.

Ideally, we want to enjoy all of the same benefits when using virtual threads with JOOQ.

**Comment by lukaseder** [comment]
> On a more serious note, virtual threads remove the need for async/reactive models

This isn't about execution models. The current API translates perfectly fine to Loom, whatever you might understand by that.

> and they fix the problem with lambdas not being able to throw checked exceptions

Lambdas have no such restriction. Some JDK functional interfaces might, but jOOQ's [`TransactionRunnable`](https://www.jooq.org/javadoc/dev/org.jooq/org/jooq/TransactionalRunnable.html#run(org.jooq.Configuration)) and [`TransactionCallable`](https://www.jooq.org/javadoc/dev/org.jooq/org/jooq/TransactionalCallable.html#run(org.jooq.Configuration)) don't. You can throw any `Throwable`

> access non-final parameters

It's rarely a real problem, certainly not one to pull the weight of this feature.

> and are much easier to read

That's just an opinion, and I'm not convinced you're right. Any procedural transaction management inevitably turns into a mess. I've done enough PL/SQL coding to know it never really works very well. Even with clever auto-rollback on `try-with-resources` statements, you still have to explicitly `commit` every time at the end of such a block. It's not easier to read/write, it's just a different style.

> debug

Your IDE can handle breakpoints in lambdas, you'll be fine

> and so on.

That's not really an argument :)

The ship has sailed in 2014 with https://github.com/jOOQ/jOOQ/issues/3229. jOOQ won't offer 2 separate transaction APIs. It's too confusing for new users or existing ones. Many people like the lambdas. Some don't. Many use Spring and ignore jOOQ's transactions. You can use the workaround I've shown. Or create a little library, it's just ~50 lines of code. Transactions are completely "orthogonal" to jOOQ, jOOQ doesn't *have to* manage them for you.

**Comment by lukaseder** [comment]
For the record, we also have `BatchedRunnable`/`BatchedCallable` as well as `ConnectionRunnable`/`ConnectionCallable`, which follow the same idiom, where a lambda contains a "unit of work", which is subject to a lifecycle (`begin transaction` / `commit`, start batch / flush batch, `acquire()` / `release()`):

- https://www.jooq.org/doc/latest/manual/sql-execution/batched-connection/
- https://www.jooq.org/doc/latest/manual/sql-execution/jdbc-connection/

It would have probably been possible to do all of these things in a try-with-resources imperative style, each time suffering from the same language flaw that the `AutoCloseable::close` method doesn't know whether there was an exception or not, so each time leaving the burden to "commit" the unit of work manually.

Implementing a procedural transaction API as an alternative to the lambda version would mean that we'd also have to implement procedural APIs for the other things (and future things) for consistency reasons. I don't like where this would be going.

**Comment by cowwoc** [comment]
No worries.

Committing in a unit of work was never really a problem because you can just `return` mid-way through the lambda. It was more of a matter of wanting to roll-back mid-way without having to throw an exception.

That said, I haven't had the need to do this for a couple of years now so I'm happy to leave things as-is :)

**Comment by lukaseder** [comment]
> It was more of a matter of wanting to roll-back mid-way without having to throw an exception.

Yeah, I get it. The lambdas (or also Spring's / Java EE's approach) are a bit weird once you do want that `ROLLBACK`. The assumption being that `ROLLBACK` is really an exceptional execution path (which it is, in probably 95% of all transactions). But sometimes it isn't.

From an 80/20 perspective, I think Spring / Java EE made the right decision (though I personally prefer lambdas over aspects) with making these things automatic.

**Comment by lukaseder** [comment]
While the suggestion of offering a transaction *API* in `DSLContext` has been rejected, adding support for transaction *statements* is still on the table and will be implemented in jOOQ 3.18, in fact: https://github.com/jOOQ/jOOQ/issues/7106

There's a subtle difference:

- Transaction *API* would mean `ctx.commit()` executes immediately, just like JDBC's `connection.commit()`
- Transaction *statements* means `ctx.commit()` returns a `Query`, on which users can call `execute()`, and which users can embed in `begin(...)` procedural blocks

The latter is much more useful in general, e.g. the parser can parse `ROLLBACK TO SAVEPOINT` statements in statement batches, and translate that to other dialects without necessarily executing the statement, all the while offering users who requested this feature here an *almost* identical experience (granted, the mandatory `execute()` call can be annoying. The `@CheckReturnValue` annotation at least helps troubleshoot forgotten `execute()` calls in some IDEs).
